### PR TITLE
expand CI runs to cover Nix and macOS

### PR DESCRIPTION
**Pull request by abathur**
Edit: Commits described in original post are mostly obsolete.

This is ~roughly one functional piece, but I've broken it down into a set of commits, backed by CI runs, to (hopefully) make it easier to evaluate. I've tried to keep it compatible with your current runs so that you can merge it directly into master, but I can also re-open against a different branch if you like.

The first 2 commits do most of the work, but avoid Oil code. The other commits focus on getting lights green. Rather than try to pre-answer every question, I'll leave a few notes, link to the CI builds, and field any questions you have.

### add CI runs for macos/nix and move tests into script  	9617b96 

[CI run at this commit](https://travis-ci.org/github/abathur/oil/builds/661386499)

- Extracts the .travis.yml test script into its own file for reuse
- adds multiple CI jobs covering Nix/macOS (and allows them to fail)
- I included nix-build jobs because of edge-case issues with in-tree nix-shell builds. I think it'll be easier to discuss the tradeoffs when you can play with them.

### refactor shell.nix and add out-of-tree build (default.nix)  	a5614f7

[CI run at this commit](https://travis-ci.org/github/abathur/oil/builds/661389133)

- Adds a default.nix (which is what nix-build runs by default)
- Adds Python a setup.py and MANIFEST.in leveraged by this build
- Refactor overlapping parts of shell/default.nix into a few shared files. (I want to clean this up further)
- Builds precise versions of test shells

### [types/run] fix PYTHONPATH for nix-shell	3650f96

[CI run after both commits above](https://travis-ci.org/github/abathur/oil/builds/661536760)

### [test/spec] use type builtin instead of which  	13148a8

[CI run after both commits above](https://travis-ci.org/github/abathur/oil/builds/661546516)

### use Nix-specified test shells when available  	896708c

[CI run at this commit](https://travis-ci.org/github/abathur/oil/builds/661559322)

### CODE REVIEW

**Comment by andychu** [comment]
I just created a domain that Travis can publish to so this will tie in nicely

http://travis-ci.oilshell.org/results/

I think the top level `test.sh` may be a good idea -- I generally dislike writing shell scripts in YAML files.  The only reason they were separate is because Travis had a nice expandable log and timing.

However now that we can publish our own dashboard, that's less important.

I created `services/travis.sh` in my branch.  I think some of the shell in YAML should go in `services/travis-nix.sh` or something like that.

----

My changes aren't that large but I am about to land something soon:

https://travis-ci.org/github/oilshell/oil/builds/661773429

**Comment by andychu** [comment]
Also note the style of 

https://github.com/oilshell/oil/blob/master/.travis.yml

invoking:

https://github.com/oilshell/oil/blob/master/services/travis.sh

and other scripts.  In general I'd like more stuff in plain shell, not shell in YAML or shell in Nix.  It makes it harder to test otherwise.

I don't really understand the Nix files still; it might time to fill out:

https://github.com/oilshell/oil/wiki/Developing-Oil-With-Nix

with some details on how people are supposed to use it, as well as how it's setup in the continuous build.

Well personally I put most comments in `services/travis.sh`.  That could be OK too.

File: .travis.yml [hunk 1]
```diff
@@ -2,58 +2,116 @@
...
-    # Running serially is slow, but easier to debug ...
-    #- test/spec.sh oil-all-serial
-    #- test/spec.sh osh-all-serial
-
-notifications:
-    on_success: change
-    on_failure: change  # `always` will be the setting once code changes slow down
+env:
+  global:
+    - secure: "tzg0cVSgJ5KzNbC2Ne9zvSTjnIe+dvc/Re3QA/7xXfH0cukSigmUQySEwAeEUw37Q3n8Z5Z1AdLnQcN13IMFCNL0azL5OGRU4tyF6FOoUIrPe9GEYfQKfQ+US3pAr4BcaOR6To+redAxVOL+CjpB8LRkZcoR2PFMIPEBxhq2R/RrCU3eAeWP1mFXkvmB+gR77iFrLAKVW3elY3Jo77Kv8EbpD6d1CIsc1aboKHXt4kYRLUSLZcCpei49VLGp7h9oZ9nMc0A2wBnpbE8uO/Mlt9o/PTA0REAt3ecWuhhJDKL0qLfU8frhhWSwY4SukpPibs/u/FGyFPZLHxx44WLYbnfkRb+vrHS87lipBhHECSRhAJbKjAEDnHhscN9qOyI5j0P21qG+NQGIUOmtbSBm9/GOatrI9dEukrXgwSay0qw9kt/DovAxvWM0TKDp9ULRHj+5kMIGtF/XMeIPhcssC1c7Odq2suCDBOnKtj+swzb018NYuM2XHBcm7ljR/GRm0CDtKYecy50w0Qgo5y2hkF+vNMoMXPVIQVktuqhWWfln8ZAy5bVPGJtTpKfVQiGaDJxCPevfoMHR4i8Zb+BwXn0q+hsf0++iamQPrNuFaWWct2JuTRAbhzP03I0XOsAcIh7QQUvLMiQdkNdlJnCOI6PBTJ87KTu+yWctLuFtseQ="
```
[the remark this answers is not in the document]
Comment (abathur) [reply .travis.yml]: Encrypted credential for cachix (a Nix binary cache). We'll need to transition it to an oil cred, I think, but it wasn't hard to set up (git login, etc.)

File: test/sh_spec.py [hunk 2]
```diff
@@ -1052,6 +1052,11 @@ def MakeTestEnv(opts):
     'PATH': opts.path_env,
     'LANG': opts.lang_env,
   }
+
+  if 'LOCALE_ARCHIVE' in os.environ:
+    # If present, this needs to make it through for nix-shell
+    env['LOCALE_ARCHIVE'] = os.environ['LOCALE_ARCHIVE']
```
[the remark this answers is not in the document]
Comment (abathur) [reply test/sh_spec.py]: Can you clarify here a little? 

It sounds a lot like you're asking for a copycat of `sh-spec-smoosh-env`, let's call it `sh-spec-nix-env`, but that pattern entails a copy of every test function we ever want to run from Nix, if we don't want it to break, and that looks a little absurd.

**Comment by andychu** [comment]
I think this will fix the PYTHONPATH problem:

https://github.com/oilshell/oil/commit/d1bbf23c10fa18cde61b045d96f43e52b177dc16

Still looking at others

File: shell.nix [hunk 3]
```diff
@@ -39,83 +38,38 @@ let
...
+  OIL_TEST_SHELL_BASH = "${shells.test_bash}/bin/bash";
+  OIL_TEST_SHELL_MKSH = "${shells.test_mksh}/bin/mksh";
+  OIL_TEST_SHELL_ZSH = "${shells.test_zsh}/bin/zsh";
+  OIL_TEST_SHELL_ASH =
+    pkgs.lib.optionalString stdenv.isLinux "${shells.test_busybox}/bin/ash";
 
   # do setup work you want to do every time you enter the shell
   # Here are a few ideas that made sense to me:
   shellHook = ''
+    export _OVM_RESOURCE_ROOT="$PWD"
```
[the remark this answers is not in the document]
Comment (abathur) [reply shell.nix]: I keep cycling through a pattern where we do this, and I remove it, and then I run into more, and lose part of my life figuring it out from scratch in a fresh context each time.

Here's the only live example I'm aware of _off the top of my head_, with the --version flag:
https://travis-ci.org/github/abathur/oil/jobs/661816072#L3332 (working)
https://travis-ci.org/github/abathur/oil/jobs/661817685#L3332 (failing)
Diff: https://github.com/abathur/oil/compare/d08779c5ee65...096e3f9c3921

**Comment by andychu** [comment]
I fixed the $PATH issues in a different way...

https://github.com/oilshell/oil/commit/84f8f51f751c86d055f5336ce38ffde315b2ac5b

Sorry it stomped on your change but I think it's cleaner.  Let me know what you think.

Basically the shell scripts like `test/spec.sh` and `build/codegen.sh` are mostly "abstract", and then they choose `build/dev-shell.sh` if we're not using `nix-shell`.

**Comment by andychu** [comment]
OK I also addressed this.

https://github.com/oilshell/oil/commit/fbfdf630f1f52c2f4651f53101e53900f7ee2e0b

So I think it should be mostly `.nix` files and the `.travis.yaml`.

I'm going to be working in `services/travis.sh` to publish spec test HTML so hopefully there shouldn't be more merging.

**Comment by andychu** [comment]
Hm or maybe I should do something like your `test.sh`, because I think that helps with publishing HTML results.  Basically it shouldn't stop at the first failure (?)

**Comment by andychu** [comment]
I think the problem with _OVM_RESOURCE_ROOT is that the error message is terrible... it should raise a different exception and show the filename or something.   But OK I guess it's fine to leave it in the Nix file for now.

File: test/sh_spec.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (abathur) [reply test/sh_spec.py]: Ok. I'll cut this per your edit, whenever I get done rebasing...

**Comment by abathur** [comment]
Gave up on rebase after git got confused and wasted 30 minutes.

Pulled the source in on a fresh branch, but I'm blocked on a dangling reference to all-for-minimal. https://travis-ci.org/github/abathur/oil/jobs/661835130#L1374

I'm out of energy. I'll revisit this tomorrow.

**Comment by andychu** [comment]
I changed that to `test/unit.sh travis` to be consistent with the other ones.

https://github.com/oilshell/oil/blob/master/.travis.yml#L46

I am going to rewrite the `script:` to move it into a shell script as mentioned.  As mentioned it will publish to travis-ci.oilshell.org so we won't rely as much on the Travis UI.

Since that is going to also change `.travis.yml` it might be better to hold off.

----

I think it might be possible to use Travis `cache:` and build and test under Alpine, e.g. `test/spec-alpine.sh`.  In case it wasn't clear, that is where this is going.  We'll have 3 or 4 different builds including the Nix one, and they will all publish to travis-ci.oilshell.org, so we can keep track of what's working where.

Sorry for the conflicts but I had been working on Travis for awhile.  I tend to be very particular about "factoring" because otherwise shell can get messy.  e.g. I like how `build/dev-shell.sh` turned out, etc.

File: shell.nix [hunk 4]
```diff
@@ -39,83 +38,32 @@ let
...
-
-  # a different option:
-  # LOCALE_ARCHIVE = lib.optionalString (stdenv.hostPlatform.libc == "glibc") "${glibcLocales}/lib/locale/locale-archive";
+  LANG = "en_US.UTF-8";
 
   # do setup work you want to do every time you enter the shell
   # Here are a few ideas that made sense to me:
   shellHook = ''
+    # Have to prefix test bash; there'll be 3+ on path
+    PATH="${shells.test_bash}/bin:$PATH"
```
Comment (abathur) [inline shell.nix]: I started to note this last night, but decided I'd see how your path changes went first. Ultimately, I was explicitly specifying test shells because there's more than one bash on path, but only one is *intended* for running tests.

I don't really expect it to cause trouble, but I want to say out loud that prepending it like this *also* means the _test_ bash is the bash used for env bash shebangs while in nix-shell.

File: build/codegen.sh [hunk 5]
```diff
@@ -24,7 +24,7 @@ source build/common.sh
 
 export PYTHONPATH='.:vendor/'
 
-if test -z "${IN_NIX_SHELL:-}"; then
+if test -z "${IN_NIX_SHELL:-}${NIX_BUILD_TOP:-}"; then
```
Comment (abathur) [inline build/codegen.sh]: I don't quite understand why, but this broke nix-build on my local macOS and not in CI (suspect a clean/dirty work dir thing).

File: shell.nix (hunk 4 above)
Comment (andychu) [reply shell.nix]: Yes that's a good point.  Well one thing we could do it rename it to `testbash` and then invoke that... should be a fairly small change.

**Comment by andychu** [comment]
As heads up I'm working on replacing the `script:` section... let me know when this is ready to look at again.  Off the top of my head here are some additional things:

- can we move most top-level files to a subdir like `nix/` ?
- can the CACHIX key be a file instead?  I used `travis encrypt-file` for `services/rsa_travis.enc`

**Comment by andychu** [comment]
And yeah moving shell out of `.travis.yml` will help too.

I like that for general cleanliness, but these changes show that .travis.yml merges poorly... so best to keep stuff out ofit.

- Also there is a comment about Ruby there that seems misplaced.  I don't understand the `<<` or `&` syntax either... is that required?

**Comment by abathur** [comment]
> As heads up I'm working on replacing the `script:` section... let me know when this is ready to look at again. Off the top of my head here are some additional things:

Metaphorically close. I wondered, while discussing the test_bash thing, if the wrapper had actually been made moot by properly passing LOCALE_ARCHIVE into the test environment. If so, I may be able to cut some of that complexity out, which I'd see as a big win.

> * can we move most top-level files to a subdir like `nix/` ?

Done (in a local commit; up when I finish the above).

> * can the CACHIX key be a file instead?  I used `travis encrypt-file` for `services/rsa_travis.enc`

Presumably. I'll try to just catch you on zulip about this.

> And yeah moving shell out of `.travis.yml` will help too. I like that for general cleanliness, but these changes show that .travis.yml merges poorly... so best to keep stuff out of it.

I'm not sure it merges terribly; we just happened to be moving in very different directions at the same time.

That said, I think it's best to have a test script that'll do the same thing as the CI run, outside of a CI context, minus anything that is just there to support the CI. (I don't care what this is named; I wanted to call it ci.sh, but that clashed with a completion test). I didn't do this purely for *always* reasons, though--just because I needed a simple way to run the standard test run from inside a nix-shell.

If you'd like to have that cake and eat it too, it's probably (in my most naive voice) not too hard to have a test script and use templating to render it back out into the YAML. (But I'm leery of having such essential steps that aren't themselves backed by tests to ensure they happen...)

> * Also there is a comment about Ruby there that seems misplaced.  

No, not misplaced. I'll see if there's a useful way to expand it. The default language for travis-ci is ruby. Oil doesn't set one, so it uses ruby. I noticed this while debugging some of this CI work. I haven't tried to measure times to see if specifying minimal/shell actually saves any meaningful time, but it seemed at least plausible that it would.

Screen Shot 2020-03-13 at 3 27 02 PM

> I don't understand the `<<` or `&` syntax either... is that required?
 
It's not required, but it enables me to build up modular config parts to keep this as DRY as possible. The `&<identifier>` is an anchor, `*<identifier>` is an alias, and `<<` merges the keys of the alias into the current position. If you ask really nicely, I'll shrug and decompile them into repetitive definitions for someone else to ensure stay in sync. I guess this could also be addressed with a templating engine.

**Comment by abathur** [comment]
Anyways. This is done, aside from the cachix key. (the builds should work, even without the cache key, albeit more slowly).

**Comment by andychu** [comment]
OK as mentioned on Zulip

- the jobs need comments 
- and they should be renamed to mention the artifact+env/platform, not just the platform.  e.g. `nix-build` or `nix-shell` doesn't make sense because you can build anything with Nix
- and remove as much shell in YAML as possible
- short docs are needed.  Imagine someone wants to build and try Oil quickly -- how do they use it?

I suggest breaking this PR up more.  I prefer to merge smaller things one at a time, e.g. the first commit was "anything that runs build/dev.sh minimal" which is easy to understand.  This is mixing in a whole bunch of stuff, and a lot of it was better done in another way ($PYTHONPATH, $PATH, LOCALE_ARCHIVE, etc.)

**Comment by andychu** [comment]
Also I think the main place where we will conflict now is `.travis.yml`.  I'm pretty close to getting  http://travis-ci.oilshell.org/ working which will change the `script:` section.  So that's another reason to split it up.

**Comment by abathur** [comment]
I've pulled cachix, added a readme, and added more comments.

**Comment by abathur** [comment]
This is caught up again.
